sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ConstantColumnVector.java

```diff
@@ -0,0 +1,264 @@
+    } else if (type instanceof StructType) {
+      StructType st = (StructType) type;
+      this.childData = new ConstantColumnVector[st.fields().length];
+      for (int i = 0; i < childData.length; ++i) {
+        this.childData[i] = new ConstantColumnVector(numRows, st.fields()[i].dataType());
+      }
+    } else if (type instanceof MapType) {
+      // 0: key, 1: value
+      MapType mapType = (MapType) type;
+      this.childData = new ConstantColumnVector[2];
+      this.childData[0] = new ConstantColumnVector(numRows, mapType.keyType());
+      this.childData[1] = new ConstantColumnVector(numRows, mapType.valueType());
+    } else if (type instanceof CalendarIntervalType) {
+      // 0: Months as Int, 1: Days as Int, 2: Microseconds as Long.
+      this.childData = new ConstantColumnVector[3];
+      this.childData[0] = new ConstantColumnVector(numRows, DataTypes.IntegerType);
+      this.childData[1] = new ConstantColumnVector(numRows, DataTypes.IntegerType);
+      this.childData[2] = new ConstantColumnVector(numRows, DataTypes.LongType);
+    } else {
+      this.childData = null;
+    }
+  }
+
+  protected boolean isArray() {
+    return type instanceof ArrayType || type instanceof BinaryType || type instanceof StringType ||
+        DecimalType.isByteArrayDecimalType(type);
+  }
+
+  @Override
+  public void close() {
+    byteArrayData = null;
+    childData = null;
+    arrayData = null;
+    mapData = null;
+  }
+
+  @Override
+  public boolean hasNull() {
+    return nullData == 1;
+  }
+
+  @Override
+  public int numNulls() {
+    return hasNull() ? numRows : 0;
+  }
+
+  @Override
+  public boolean isNullAt(int rowId) {
+    return nullData == 1;
+  }
+
+  public void putNull() {
+    nullData = (byte) 1;
+  }
+
+  public void putNotNull() {
+    nullData = (byte) 0;
+  }
+
+  @Override
+  public boolean getBoolean(int rowId) {
+    return byteData == 1;
+  }
+
+  public void putBoolean(boolean value) {
+    byteData = (byte) ((value) ? 1 : 0);
+  }
+
+  @Override
+  public byte getByte(int rowId) {
+    return byteData;
+  }
+
+  public void putByte(byte value) {
+    byteData = value;
+  }
+
+  @Override
+  public short getShort(int rowId) {
+    return shortData;
+  }
+
+  public void putShort(short value) {
+    shortData = value;
+  }
+
+  @Override
+  public int getInt(int rowId) {
+    return intData;
+  }
+
+  public void putInt(int value) {
+    intData = value;
+  }
+
+  @Override
+  public long getLong(int rowId) {
+    return longData;
+  }
+
+  public void putLong(long value) {
+    longData = value;
+  }
+
+  @Override
+  public float getFloat(int rowId) {
+    return floatData;
+  }
+
+  public void putFloat(float value) {
+    floatData = value;
+  }
+
+  @Override
+  public double getDouble(int rowId) {
+    return doubleData;
+  }
+
+  public void putDouble(double value) {
+    doubleData = value;
+  }
+
+  @Override
+  public ColumnarArray getArray(int rowId) {
```

Comment:
Not sure if this can work properly. Looking at `ColumnarArray`, in some cases `offset` is required from underlying `ColumnVector`, for instance, `copy`, `toBooleanArray`, etc.